subsys/net/l2/ethernet/dsa/dsa.c

```diff
@@ -0,0 +1,202 @@
+/** @file
+ * @brief DSA related functions
+ */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_dsa, CONFIG_NET_DSA_LOG_LEVEL);
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include <net/net_core.h>
+#include <net/ethernet.h>
+#include <net/net_mgmt.h>
+#include <net/dsa.h>
+
+/*
+ * Helper functions
+ */
+struct dsa_context *dsa_get_context(struct net_if *iface)
+{
+	struct ethernet_context *eth_ctx;
+
+	eth_ctx = net_if_l2_data(iface);
+	if (eth_ctx == NULL) {
+		LOG_ERR("Iface %p context not available!", iface);
+		return NULL;
+	}
+
+	return eth_ctx->dsa_ctx;
+}
+
+/*
+ * Store, in the ethernet_context for master interface, the original
+ * eth_tx() function, which will send packet with tag appended.
+ */
+int dsa_register_master_tx(struct net_if *iface, dsa_send_t fn)
+{
+	struct ethernet_context *ctx = net_if_l2_data(iface);
+
+	ctx->dsa_send = fn;
+	return 0;
+}
+
+#ifdef CONFIG_NET_L2_ETHERNET
+bool dsa_is_port_master(struct net_if *iface)
+{
+	/* First check if iface points to ETH interface */
+	if (net_if_l2(iface) == &NET_L2_GET_NAME(ETHERNET)) {
+		/* Check its capabilities */
+		if (net_eth_get_hw_capabilities(iface) &
+		    ETHERNET_DSA_MASTER_PORT) {
+			return true;
+		}
+	}
+
+	return false;
+}
+#else
+bool dsa_is_port_master(struct net_if *iface)
+{
+	return false;
+}
+#endif
+
+/*
+ * RECEIVE HANDLING CODE - ingress (ETH -> DSA slave ports)
+ */
+
+static int dsa_check_iface(struct net_if *iface)
+{
+	if (net_if_l2(iface) != &NET_L2_GET_NAME(ETHERNET)) {
+		return -ENOENT;
+	}
+
+	if (!((net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_MASTER_PORT) ||
+	    (net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_SLAVE_PORT))) {
+		return -ESRCH;
+	}
+
+	return 0;
+}
+
+
+int dsa_register_recv_callback(struct net_if *iface, net_dsa_recv_cb_t cb)
+{
+	struct ethernet_context *ctx;
+	int ret;
+
+	ret = dsa_check_iface(iface);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (cb) {
+		ctx = net_if_l2_data(iface);
+		ctx->cb = cb;
+	}
+
+	return 0;
+}
+
+int net_dsa_recv(struct net_if *iface, struct net_pkt *pkt)
+{
+	struct ethernet_context *ctx;
+	int ret;
+
+	ret = dsa_check_iface(iface);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ctx = net_if_l2_data(iface);
+	if (ctx->cb) {
+		return ctx->cb(iface, pkt);
+	}
+
+	return 0;
+}
+
+struct net_if *dsa_recv_set_iface(struct net_if *iface, struct net_pkt **pkt)
```

Comment:
So basically this "hijacks" the packet. I think we should leave `net_recv_data()` as is and you could add all the DSA specific stuff to `ethernet.c`.

Reply:
I'm not sure what do you mean by packet "hijacks" . The `dsa_recv_set_iface()` is responsible for:
1. Use `dsa_ksz8794_get_iface()` to read the tag and return proper `struct net_if` pointer to LAN{123} interfaces. That is why we need to call it early to have the tag data (from switch IC) not removed and packet interface set before it is used by network stack. In that way the packet shall be passed to proper LAN{123} interface.
2. (optionally) - The `net_dsa_recv()` function calls the user application registered callback (in the ./samples/net/dsa case it is `dsa_ll_addr_switch_cb()` which just checks the `lladst.addr`).

Please also note that in the end of `net_recv_data()` (line:427) we call `net_pkt_set_iface(pkt, iface);` which assigns iface to packet. After that the packet is passed to proper RX queues.

I could add the `dsa_recv_set_iface()` to `ethernet_recv()` but IMHO considering the above arguments the current placement of this function is optimal.

Reply:
I did quick experiment and moved the stuff from net_core.c to mcux driver. Why would this not work?
```
diff --git a/drivers/ethernet/eth_mcux.c b/drivers/ethernet/eth_mcux.c
index 8d13bfa1b8..2b85278a67 100644
--- a/drivers/ethernet/eth_mcux.c
+++ b/drivers/ethernet/eth_mcux.c
@@ -710,6 +710,7 @@ static void eth_rx(struct eth_context *context)
 {
 	uint16_t vlan_tag = NET_VLAN_TAG_UNSPEC;
 	uint32_t frame_length = 0U;
+	struct net_if *iface;
 	struct net_pkt *pkt;
 	status_t status;
 	unsigned int imask;
@@ -811,7 +812,13 @@ static void eth_rx(struct eth_context *context)
 
 	irq_unlock(imask);
 
-	if (net_recv_data(get_iface(context, vlan_tag), pkt) < 0) {
+	iface = get_iface(context, vlan_tag);
+
+#if IS_ENABLED(CONFIG_NET_DSA)
+	iface = dsa_recv_set_iface(iface, &pkt);
+#endif
+
+	if (net_recv_data(iface, pkt) < 0) {
 		net_pkt_unref(pkt);
 		goto error;
 	}
diff --git a/subsys/net/ip/net_core.c b/subsys/net/ip/net_core.c
index c0a63c7172..f60bec1fad 100644
--- a/subsys/net/ip/net_core.c
+++ b/subsys/net/ip/net_core.c
@@ -396,9 +396,6 @@ int net_recv_data(struct net_if *iface, struct net_pkt *pkt)
 	net_pkt_set_overwrite(pkt, true);
 	net_pkt_cursor_init(pkt);
 
-#if IS_ENABLED(CONFIG_NET_DSA)
-	iface = dsa_recv_set_iface(iface, &pkt);
-#endif
 	NET_DBG("prio %d iface %p pkt %p len %zu", net_pkt_priority(pkt),
 		iface, pkt, net_pkt_get_len(pkt));
```

Reply:
This will not work as you can pass `iface == NULL` to `dsa_recv_set_iface()` as the argument. This function is not checking it as such checks are already done in `net_recv_data()`. 
As a result - some allocation errors pops up and the stack is not usable anymore.

Moreover, the `net_pkt_set_overwrite(pkt, true);` is executed `net_recv_data()` as well, so packet manipulation is possible (to remove tag).